shell/platform/android/io/flutter/view/AccessibilityViewEmbedder.java

```diff
@@ -470,32 +496,102 @@ private Long getSourceNodeId(@NonNull AccessibilityNodeInfo node) {
+                // (current min is 16) because they get compiled to the common but new super
+                // type ReflectiveOperationException. As a workaround either create individual
+                // catch statements, or catch Exception. [NewApi]
+                } catch (IllegalAccessException e) {
+                    Log.w(TAG, e);
+                } catch (InvocationTargetException e) {
+                    Log.w(TAG, e);
+                }
+            } else {
+                try {
+                    return (long) longArrayGetIndex.invoke(childNodesField.get(node), child);
+                // Using identical separate catch blocks to comply with the following lint:
+                // Error: Multi-catch with these reflection exceptions requires API level 19
+                // (current min is 16) because they get compiled to the common but new super
+                // type ReflectiveOperationException. As a workaround either create individual
+                // catch statements, or catch Exception. [NewApi]
+                } catch (IllegalAccessException e) {
+                    Log.w(TAG, e);
+                } catch (InvocationTargetException | ArrayIndexOutOfBoundsException e) {
+                    Log.w(TAG, e);
+                }
             }
             return null;
         }
 
         @Nullable
         private Long getParentNodeId(@NonNull AccessibilityNodeInfo node) {
-            if (getParentNodeId == null) {
+            if (getParentNodeId != null) {
+                try {
+                    return (long) getParentNodeId.invoke(node);
+                // Using identical separate catch blocks to comply with the following lint:
+                // Error: Multi-catch with these reflection exceptions requires API level 19
+                // (current min is 16) because they get compiled to the common but new super
+                // type ReflectiveOperationException. As a workaround either create individual
+                // catch statements, or catch Exception. [NewApi]
+                } catch (IllegalAccessException e) {
+                    Log.w(TAG, e);
+                } catch (InvocationTargetException e) {
+                    Log.w(TAG, e);
+                }
+            }
+
+            // Fall back on reading the ID from a serialized data if we absolutely have to.
+            return yoinkParentIdFromParcel(node);
+        }
+
+        @Nullable
+        private static Long yoinkParentIdFromParcel(AccessibilityNodeInfo node) {
+            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+                Log.w(TAG, "Unexpected Android version. Unable to find the parent ID.");
                 return null;
             }
-            try {
-                return (long) getParentNodeId.invoke(node);
-            } catch (IllegalAccessException e) {
-                Log.w(TAG, e);
-            } catch (InvocationTargetException e) {
-                Log.w(TAG, e);
+
```

Comment:
Should we be explicit about an upper Android version bound here as we can't guarantee that the parcel structure won't change?

If R will include a public API that helps us it we should definitely do it, if not I'm not sure if we want to require manual intervention here for any new Android version, or only require it if it breaks(which can potentially lead to more difficult debugging).

Not sure what's better, what are your thoughts?

Reply:
Yeah I was going back and forth on this too. Reason I just lower bounded it is that I'm not sure that we would catch the version code bump in time, since this isn't being tested. (I am digging into how to integration test this in `flutter/flutter` now though, which would probably fix that.) But as it is now if this auto failed on higher Android versions it wouldn't come through in an obvious way back to us, other than us really knowing we needed to check it out. So it would be a guaranteed loss without a good way to prevent it. Maybe a TODO here to bound this once we have CI pointed at this?

Reply:
sgtm